File: JobManager/Models/Message.swift
//
//  Message.swift
//  HASLIb
//
//  Created by Abel Gancsos on 5/3/20.
//  Copyright © 2020 Abel Gancsos. All rights reserved.
//

import Foundation
import Common

public class Message : Item {
    private var messageId       : Int;
    private var text            : String;
    private var level           : TraceLevel;
    private var category        : TraceCategory;
    private var lastUpdatedDate : String;
    
    public init(id : Int = -1) {
        self.messageId = id;
        self.text = "";
        self.level = .NONE;
        self.category = .NONE;
        self.lastUpdatedDate = "";
    }
    
    public init(json : String) {
        let jsonObject : JsonObject = JsonObject(json: json);
        self.messageId = Int(jsonObject.getKey(key: "id") as! String) ?? -1;
        self.text = jsonObject.getKey(key: "text") as! String;
        self.level = TraceLevel.init(rawValue: Int(jsonObject.getKey(key: "level") as! String) ?? 0) ?? .NONE;
        self.category = TraceCategory.init(rawValue: Int(jsonObject.getKey(key: "category") as! String) ?? 0) ?? .NONE;
        self.lastUpdatedDate = jsonObject.getKey(key: "lastUpdatedDate") as! String;
    }
    
    public func toJsonString() -> String {
        var result = "{";
        result += "\"id\":\"\(self.messageId)\",";
        result += "\"level\":\"\(self.level.rawValue)\",";
        result += "\"category\":\"\(self.category.rawValue)\",";
        result += "\"text\":\"\(self.text)\",";
        result += "\"lastUpdatedDate\":\"\(self.lastUpdatedDate)\"";
        result += "}";
        return result;
    }
    
    public func getId() -> Int { return self.messageId; }
    public func getText() -> String { return self.text; }
    public func getLevel() -> TraceLevel { return self.level; }
    public func getCategory() -> TraceCategory { return self.category; }
    public func getLastUpdatedDate() -> String { return self.lastUpdatedDate; }
    public func setText(a : String) { self.text = a; }
    public func setLevel(a : TraceLevel) { self.level = a; }
    public func setCategory(a : TraceCategory) { self.category = a; }
    public func setLastUpdatedDate(a : String) { self.lastUpdatedDate = a; }
}

File: jmserver/src/classes/jmserver/models/Message.go
package models
import (
    "../../common"
    "encoding/json"
)

type Message struct {
	id              int    `json:id`;
	text            string `json:text`;
    level           int    `json:level`;
    category        int    `json:category`;
    lastUpdatedDate string `json:lastUpdatedDate`;
}
func (x *Message) ReloadFromJson(json string) {
    var dict = common.StrToDictionary([]byte(json));
    x.id = dict["id"].(int);
    x.text = dict["text"].(string);
    x.level = dict["level"].(int);
    x.category = dict["category"].(int);
    x.lastUpdatedDate = dict["lastUpdatedDate"].(string);
}

func (x *Message) ToJsonString() string {
    rawJson, _ := json.Marshal(x);
    return common.DictionaryToJsonString(common.StrToDictionary(rawJson));
}

func (x Message) ID() int { return x.id; }
func (x Message) Text() string { return x.text; }
func (x Message) Level() int { return x.level; }
func (x Message) Category() int { return x.category; }
func (x Message) LastUpdatedDate() string { return x.lastUpdatedDate; }
func (x *Message) SetID(id int) { x.id = id; }
func (x *Message) SetText(text string) { x.text = text; }
func (x *Message) SetLevel(level int) { x.level = level; }
func (x *Message) SetCategory(category int) { x.category = category; }
func (x *Message) SetLastUpdatedDate(ud string) { x.lastUpdatedDate = ud; }
